### PR TITLE
Allow Kind tests to mount `~/.m2/`

### PR DESCRIPTION
**Pull request by jglick**
I have not had success using the various profiles and `connectorHost` / `jenkins.host.address` with Kind. Running tests inside the cluster is more robust. Unfortunately it is also slow to start up, due to the need to download lots of stuff to a volume, and you cannot share local snapshots. The system in this PR ([docs](https://kind.sigs.k8s.io/docs/user/configuration/#extra-mounts)) seems to work much better to test e.g. #1494 using

```diff
diff --git pom.xml pom.xml
index df3627bc..0b178e0e 100644
--- pom.xml
+++ pom.xml
@@ -62,7 +62,7 @@
       <dependency>
         <groupId>io.jenkins.tools.bom</groupId>
         <artifactId>bom-${bom}</artifactId>
-        <version>${bom.version}</version>
+        <version>2675.v1515e14da_7a_6</version>
         <type>pom</type>
         <scope>import</scope>
       </dependency>
@@ -134,6 +134,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
       <artifactId>workflow-cps</artifactId>
+      <version>999999-SNAPSHOT</version>
       <optional>true</optional>
     </dependency>
     <dependency>
@@ -152,6 +153,7 @@
       <!-- DeclarativeAgent -->
       <groupId>org.jenkinsci.plugins</groupId>
       <artifactId>pipeline-model-extensions</artifactId>
+      <version>999999-SNAPSHOT</version>
       <optional>true</optional>
     </dependency>
     <dependency>
```

to check things like

```diff
diff --git plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovySourceFileAllowlist.java plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovySourceFileAllowlist.java
index 64c13495..987d86a9 100644
--- plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovySourceFileAllowlist.java
+++ plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovySourceFileAllowlist.java
@@ -138,9 +138,11 @@ public abstract class GroovySourceFileAllowlist implements ExtensionPoint {
             String urlString = url.toString();
             for (GroovySourceFileAllowlist allowlist : GroovySourceFileAllowlist.all()) {
                 if (allowlist.isAllowed(urlString)) {
+                    LOGGER.info(() -> "TODO allowing " + urlString + " based on " + allowlist);
                     return true;
                 }
             }
+            LOGGER.info(() -> "TODO denying " + urlString);
             return false;
         }
 ```

```diff
diff --git pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
index 4a29f937..c23b8618 100644
--- pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
+++ pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
@@ -38,6 +38,7 @@ import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.logging.Logger;
 
 /**
  * Descriptor for {@link WithScriptDescribable}.
@@ -46,8 +47,11 @@ import java.util.Set;
  */
 public abstract class WithScriptDescriptor<T extends WithScriptDescribable<T>> extends Descriptor<T> {
 
+    private static final Logger LOGGER = Logger.getLogger(WithScriptDescriptor.class.getName());
+
     protected WithScriptDescriptor() {
         ExtensionList.lookupSingleton(WithScriptAllowlist.class).registerScript(getScriptResource());
+        LOGGER.info(() -> "TODO registering " + getScriptResource() + " from " + this);
     }
 
     /**
```

Holding in the draft for the moment not just because I am unsure whether this will actually work on a CI machine, but also because it would be nice to retain the ability to use `test-in-k8s.sh` with non-Kind clusters.

### CODE REVIEW

**Comment by jglick** [comment]
> whether this will actually work on a CI machine

It does. But it is not any faster than trunk, because each `kind-*` branch runs in its own VM with no access to a cached local repo. The revised system _is_ much faster locally on a dev machine, though you need to manually run https://github.com/jenkinsci/kubernetes-plugin/blob/62116ecf707dbb0b0421c5093c0352a66f110a6f/kind.sh#L18-L28 locally to set up (plain `kind create cluster` will not work). That could be extracted to a helper script.

> it would be nice to retain the ability to use `test-in-k8s.sh` with non-Kind clusters

Not sure how to address that without maintaining two copies of at least https://github.com/jenkinsci/kubernetes-plugin/blob/62116ecf707dbb0b0421c5093c0352a66f110a6f/test-in-k8s.yaml#L6-L32 which I guess would complicate https://github.com/jenkinsci/kubernetes-plugin/blob/62116ecf707dbb0b0421c5093c0352a66f110a6f/.github/renovate.json#L29-L33 or something. Maybe [Kustomize](https://kustomize.io/) can handle this? Or some `yq` hacks?

@Vlatombe WDYT—is this worth polishing up?

File: README.md [hunk 1]
```diff
@@ -1034,7 +1052,7 @@ system property to the (host-only or NAT) IP of your host:
 
     mvn clean install -Djenkins.host.address=192.168.99.1
 
-### Integration Tests with Microk8s
+### Integration tests with Microk8s
```
Comment (Vlatombe) [inline README.md:1055]: Do we want to keep Microk8s? Nowadays I only use kind.

**Review by Vlatombe — APPROVED** [review]

File: kind-preload.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+set -euxo pipefail
+cd $(dirname $0)
+
+PRE_LOAD_IMAGES=()
+PRE_LOAD_IMAGES+=($(grep -e image: test-in-k8s.yaml | cut -d ':' -f 2- | xargs))
+PRE_LOAD_IMAGES+=($(grep -h --include="*.groovy" -e "^\s*image: .*$" -R src/test/resources | sed -e "s/^[[:space:]]*image: //" | sort | uniq | grep -v "windows" | grep -v "nonexistent" | grep -v "invalid" | xargs))
+PRE_LOAD_IMAGES+=($(grep -e FROM src/main/resources/org/csanchez/jenkins/plugins/kubernetes/Dockerfile | cut -d ' ' -f 2-))
+if [[ -v cluster ]]
+then
+  name_arg='--name $cluster'
```
Comment (Vlatombe) [inline kind-preload.sh:11]:
```suggestion
  name_arg="--name $cluster"
```

**Comment by jglick** [comment]
`KubernetesPipelineTest.dynamicPVCWorkspaceVolume` failure :eyes:

**Comment by jglick** [comment]
Could not reproduce the test failure locally even without using the new M2 mounting stuff (which CI does not now use). Can see if https://github.com/jenkinsci/kubernetes-plugin/pull/1550#discussion_r1599973932 helps…

**Pull request merged**